### PR TITLE
[Utils] Fix windows utf8 paths

### PR DESCRIPTION
**Pull request by Warrows**
Since introducing a change in leveldb using utf8 some setups using windows and paths with non-ANSI chars were not able to open the wallet anymore. We fix the situation by backporting the work toward this end fro bitcoin. Tested in an actual environement.

Also we refactor a bit the wallet database interaction.

Closes https://github.com/PIVX-Project/PIVX/issues/940
Backports most changes from https://github.com/bitcoin/bitcoin/issues/13869

### CODE REVIEW

File: src/wallet/wallet.cpp [hunk 1]
```diff
@@ -426,6 +427,45 @@ std::set<uint256> CWallet::GetConflicts(const uint256& txid) const
...
+{
+    if (GetBoolArg("-disablewallet", false))
+        return true;
+
+    uiInterface.InitMessage(_("Verifying wallet..."));
+
+    std::string strError;
+    if (!CWalletDB::VerifyEnvironment(walletFile, GetDataDir().string(), strError))
+        errorString += strError;
+        return true;
```
Comment (random-zebra) [inline src/wallet/wallet.cpp]: missing `if` braces `{}`. Everything after this `return true;` is unreachable code.
Plus, it should return false.

File: src/wallet/wallet.cpp [hunk 2]
```diff
@@ -426,6 +427,45 @@ std::set<uint256> CWallet::GetConflicts(const uint256& txid) const
...
+        CWallet dummyWallet;
+        if (!CWalletDB::Recover(walletFile, (void *)&dummyWallet, CWalletDB::RecoverKeysOnlyFilter))
+            return false;
+    }
+
+    std::string strWarning;
+    bool dbV = CWalletDB::VerifyDatabaseFile(walletFile, GetDataDir().string(), strWarning, strError);
+    if (!strWarning.empty())
+        warningString += strWarning;
+        return true;
```
Comment (random-zebra) [inline src/wallet/wallet.cpp]: same here. no braces, and misleading indentation. Also why return true?

File: src/util.cpp [hunk 3]
```diff
@@ -256,9 +259,12 @@ int LogPrintStr(const std::string& str)
...
             fReopenDebugLog = false;
-            boost::filesystem::path pathDebug = GetDataDir() / "debug.log";
-            if (freopen(pathDebug.string().c_str(), "a", fileout) != NULL)
-                setbuf(fileout, NULL); // unbuffered
+            fs::path pathDebug = GetDataDir() / "debug.log";
+            FILE* new_fileout = fsbridge::fopen(pathDebug, "a");
+            if (new_fileout)
+                setbuf(new_fileout, nullptr); // unbuffered
+                fclose(fileout);
+                fileout = new_fileout;
```
Comment (random-zebra) [inline src/util.cpp]: same here. no braces, and misleading indentation.

File: src/wallet/wallet.cpp (hunk 1 above)
Comment (Warrows) [reply src/wallet/wallet.cpp]: Good catch. I recall how and when I messed this up. It's an overlook from backporting https://github.com/bitcoin/bitcoin/pull/8574 which uses the oneliner `return InitError(strError);`

File: src/pivx-cli.cpp [hunk 4]
```diff
@@ -84,7 +85,7 @@ static bool AppInitRPC(int argc, char* argv[])
         fprintf(stdout, "%s", strUsage.c_str());
         return false;
     }
-    if (!boost::filesystem::is_directory(GetDataDir(false))) {
+    if (!fs::is_directory(GetDataDir(false))) {
```
Comment (flo071) [inline src/pivx-cli.cpp:88]: fs: not declared -> #include "fs.h" missing?

**Comment by random-zebra** [comment]
Should we close this one? Quite old and stale.
Plus it's a big change for such a small issue. 
Imo we better backport the PRs listed in bitcoin/bitcoin#13869, in multiple batches, later, when our code is more mature.

**Comment by Warrows** [comment]
Agreed

**Pull request closed without merging**